### PR TITLE
Add edits in the locked state

### LINKED ISSUE #3492: Expenditures: As a user I can make edits in the locked state
[Provided by Michał via Contractors project management platform]

## Description
The user should be able to edit a locked expenditure, but clicking edit, the state of the expenditure will revert to an editable state. There should also be a modal that appears when the user confirms the changes. This will create a motion for the changes.

## Acceptance Criteria
 - The user is able to go from a locked state to an editable state
 - The user is presented with a modal to confirm their changes.
 - The UI shows the newly created linked motion.

## Design

[Figma Link - Showing edit icon](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=2416%3A313415)
[Figma Link - Showing edit state](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=2416%3A314175)
[Figma Link - Showing confirm edit modal](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=2416%3A314175)
[Figma Link - Showing linked edit motions](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=2420%3A1074809)

Recipient - hover

Recipient - hover3

Recipient - hover4

Screenshot 2022-06-14 133113

### PR DESCRIPTION
**Pull request by agnieszkajarosikloj**
## Description

This PR adds ability to edit a locked expenditure, after clicking edit icon. The state of the expenditure is revert to an editable. Also adds a modal where user can confirm the changes and displays newly created linked motion. UI only.
[Figma Link - Showing edit icon](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=2416%3A313415)
[Figma Link - Showing edit state](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=2416%3A314175)
[Figma Link - Showing confirm edit modal](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=2416%3A314175)
[Figma Link - Showing linked edit motions](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=2420%3A1074809)

**New stuff** ✨

* Edit button that reverts to an editable state
Screenshot 2022-07-14 at 14 48 48
* Modal where user can confirm changes
Screenshot 2022-07-14 at 15 20 52

**Changes** 🏗

* The Stages component has been changed to separate the Stages inside the form from the one with locked values. It has been done to avoid calling the 'useFormikContext' hook outside of formik context.
* StakeExpenditureDialog has been changed - used only Toggle with correct props
Screenshot 2022-07-27 at 18 41 33

Resolves #3492

### CODE REVIEW

File: src/modules/dashboard/components/Dialogs/EditExpenditureDialog/EditExpenditureDialog.tsx [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+import EditExpenditureDialogForm from './EditExpenditureDialogForm';
+
+export const MSG = defineMessages({
+  header: {
+    id: 'dashboard.EditConfirmDialog.header',
+    defaultMessage: 'Create a motion to edit payment',
+  },
+  headerForce: {
+    id: 'dashboard.EditConfirmDialog.headerForce',
+    defaultMessage: 'Create a motion to edit expenditure',
```
Comment (arrenv) [inline src/modules/dashboard/components/Dialogs/EditExpenditureDialog/EditExpenditureDialog.tsx:17]: We can combine these and have them as "Create a motion to edit payment"

File: src/modules/dashboard/components/Dialogs/EditExpenditureDialog/EditExpenditureDialog.tsx [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    id: 'dashboard.EditConfirmDialog.descriptionLabel',
+    defaultMessage: `Explain why you're changing the payment (optional)`,
+  },
+  cancelText: {
+    id: 'dashboard.EditConfirmDialog.cancelText',
+    defaultMessage: 'Back',
+  },
+  confirmText: {
+    id: 'dashboard.EditConfirmDialog.confirmText',
+    defaultMessage: 'Create motion',
```
Comment (arrenv) [inline src/modules/dashboard/components/Dialogs/EditExpenditureDialog/EditExpenditureDialog.tsx:52]: Can we make it "Create Motion" with the capital "M"

**Review by arrenv — CHANGES_REQUESTED** [review]
Thank you @agnieszkajarosikloj 

I am getting a unique key error in the console.

0181a02f-da39-4c41-bd2c-54f5cc9626be

I am able to Update even if no values were changed, I think it would be better to disable the Update button until values are changed. Or, show a value in the dialog that no values have been changed in the same `FormSection` Element

c202cd25-87c3-45b2-bf62-5a55f5ab9842

The token and the value in the `FormSection` are not the same font size, and is not vertically aligned.

After clicking discard with the last element, you get the same empty dialog issue.

c887ad57-5c5c-4140-bdc1-782fd32ef3c0


c202cd25-87c3-45b2-bf62-5a55f5ab9842

**Review by willm30 — CHANGES_REQUESTED** [review]
@agnieszkajarosikloj  Made a few comments below 🙂

1. Alignment here looks off when changing amount from one coin to two (e.g. not centered vertically).
new-amount

1. When toggling force, I would say the warning should have some bottom padding.
annotation-padding

1. When changing the delay, per the design it would be "1h" which would also remove the issue with the "hours" being plural here. 
hours

1. There's too much space beneath the first row compared with the design, and also, I think the letter spacing of the labels in the design is greater. I believe the colour is also a slightly lighter grey.
spacing

1. I know it's not strictly related to this issue, but "F. ex." should be replaced with "E.g."
fex
 
1. The label should be in the center of the icon. If you set the \<i\> tag's display to `display: flex` it solves it.
payments-vertical
 
1. The design implies the ability to change owner, however there is no user picker dropdown at the owner section when you click edit. 

change-owner

owner-picker

**Comment by willm30** [comment]
@agnieszkajarosikloj A question for you: when I edit values and confirm, they aren't updating in the list in the same "dropdown" style as the original. In the screenshots below I added two new recipients but can't see what I added. Is that in scope for this PR?
edit-locked

**Comment by agnieszkajarosikloj** [comment]
> @agnieszkajarosikloj A question for you: when I edit values and confirm, they aren't updating in the list in the same "dropdown" style as the original. In the screenshots below I added two new recipients but can't see what I added. Is that in scope for this PR? edit-locked

I added a temporary function that needs to be replaced with a call to backend. Now, if you "force" edition, you'll see your changes immediately. When you create a motion to change expenditure, you'll see changes after setTimeout.

**Comment by willm30** [comment]
> I added a temporary function that needs to be replaced with a call to backend. Now, if you "force" edition, you'll see your changes immediately. When you create a motion to change expenditure, you'll see changes after setTimeout.

@agnieszkajarosikloj Ok thanks for clarifying. A couple more notes:

1. When you "force" an update, it tells you an annotation is required but lets you submit the form without one
force-payment

2. I have noticed elsewhere in the project the colour of the "bin" icon when deleting a recipient is red but here it's grey. Mentioning just in case.

Approving in advance as is working well for me.

**Review by willm30 — APPROVED** [review]

**Review by Julianahlk — CHANGES_REQUESTED** [review]
Good work so far! A few changes from my end:

1. I noticed the completed button color in this PR is blue which differs from other PRs so for consistency can we keep it either all blue or grey?  
Screen Shot 2022-07-29 at 7 46 36 PM

2. In the example below, the button completed should not be showing up since there is a claim delay of 1 hour which doesn't match the actual state or eligibility to claim. 
Screen Shot 2022-07-29 at 7 48 30 PM

I was expecting the orange button show up instead and the claim button to be in its off state.
Screen Shot 2022-07-29 at 8 03 58 PM

3. Value is supposed to be aligned to the top or bottom?
Screen Shot 2022-07-29 at 7 57 19 PM
Screen Shot 2022-07-29 at 7 58 15 PM

4. Can we align the following text to the left: "Recipient has been deleted"
Screen Shot 2022-07-29 at 8 13 42 PM
5. Last thing, I noticed this orange dot doesn't have a tooltip 
Screen Shot 2022-07-29 at 8 16 58 PM
And the blue dots do 
Screen Shot 2022-07-29 at 8 17 17 PM

**Comment by agnieszkajarosikloj** [comment]
> Good work so far! A few changes from my end:
> 
> 1. I noticed the completed button color in this PR is blue which differs from other PRs so for consistency can we keep it either all blue or grey?
> 
> Screen Shot 2022-07-29 at 7 46 36 PM
> 
> 2. In the example below, the button completed should not be showing up since there is a claim delay of 1 hour which doesn't match the actual state or eligibility to claim.
> 
> Screen Shot 2022-07-29 at 7 48 30 PM
> 
> I was expecting the orange button show up instead and the claim button to be in its off state. Screen Shot 2022-07-29 at 8 03 58 PM
> 
> 3. Value is supposed to be aligned to the top or bottom?
> 
> Screen Shot 2022-07-29 at 7 57 19 PM
> 
> Screen Shot 2022-07-29 at 7 58 15 PM
> 
> 4. Can we align the following text to the left: "Recipient has been deleted"
> 
> Screen Shot 2022-07-29 at 8 13 42 PM
> 
> 5. Last thing, I noticed this orange dot doesn't have a tooltip
> Screen Shot 2022-07-29 at 8 16 58 PM
> 
> And the blue dots do
> Screen Shot 2022-07-29 at 8 17 17 PM
> 
> Thank you!

@Julianahlk thank you for comments.
1. I've fixed 'Completed' button color in this PR: https://github.com/JoinColony/colonyDapp/pull/3527/files#diff-692e723973e10e96eea291e621374752d4483680460c120175c41afed277eef3
I don't want to copy the same code here.
2. It is similar to the first point. Claim delay [PR](https://github.com/JoinColony/colonyDapp/pull/3527) isn't merged yet to the [advanced-expenditures](https://github.com/JoinColony/colonyDapp/tree/feature/advanced-expenditures) branch, so delay doesn't work yet.
3. and 4. I've added fixes. Thank you for pointing this out.
5. I couldn't find tooltip for this dot in the desing. @arrenv should tooltip be here?  If so, can you please tell me what should be displayed here?

**Comment by Julianahlk** [comment]
> 1. I've fixed 'Completed' button color in this PR: https://github.com/JoinColony/colonyDapp/pull/3527/files#diff-692e723973e10e96eea291e621374752d4483680460c120175c41afed277eef3
>    I don't want to copy the same code here.
> 2. It is similar to the first point. Claim delay [PR](https://github.com/JoinColony/colonyDapp/pull/3527) isn't merged yet to the [advanced-expenditures](https://github.com/JoinColony/colonyDapp/tree/feature/advanced-expenditures) branch, so delay doesn't work yet.
> 3. and 4. I've added fixes. Thank you for pointing this out.
> 4. I couldn't find tooltip for this dot in the desing. @arrenv should tooltip be here?  If so, can you please tell me what should be displayed here?

1. I see the changes now. Thank you!
2. Got it!
3. Happy to help!

File: src/modules/dashboard/components/ExpenditurePage/Stages/LockedStages.tsx [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+import styles from './Stages.css';
+
+const MSG = defineMessages({
+  motion: {
+    id: 'dashboard.ExpenditurePage.Stages.motion',
+    defaultMessage: `There is an active motion for this expenditure`,
+  },
+});
+
+const displayName = 'dashboard.ExpenditurePage.LockedStages';
```
Comment (chinins) [inline src/modules/dashboard/components/ExpenditurePage/Stages/LockedStages.tsx:18]: missing Stages

File: src/modules/dashboard/components/ExpenditurePage/Stages/LockedStages.tsx [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+import Tag from '~core/Tag';
+import { Motion, State } from '~pages/ExpenditurePage/ExpenditurePage';
+
+import { MotionStatus, Status } from './constants';
+import Stages from './Stages';
+import styles from './Stages.css';
+
+const MSG = defineMessages({
+  motion: {
+    id: 'dashboard.ExpenditurePage.Stages.motion',
```
Comment (chinins) [inline src/modules/dashboard/components/ExpenditurePage/Stages/LockedStages.tsx:13]: missing LockedStages

File: src/modules/dashboard/components/ExpenditurePage/Stages/Stages.tsx [hunk 5]
```diff
@@ -115,8 +109,46 @@ const Stages = ({ states, activeStateId }: Props) => {
...
+      const role = ColonyRole.Arbitration;
+
+      if (status === Status.ForceEdited && index === activeIndex) {
+        return (
+          <div className={styles.labelComponent}>
+            {typeof label === 'object' && label?.id ? (
+              <FormattedMessage {...label} />
+            ) : (
+              label
+            )}{' '}
```
Comment (chinins) [inline src/modules/dashboard/components/ExpenditurePage/Stages/Stages.tsx:130]: Needs to be combined in a translation string and used with FormattedMessage. Not good to have spaces like this if possible

**Review by chinins — CHANGES_REQUESTED** [review]
Nice work - a lot of it :)

This PR is very large and difficult to review. PRs should be smaller. If you fee that it's getting large (like this one), please, talk to Arren to split the issue.

In general, I would prefer for you to split the code in more digestible parts (make more subcomponents). I have commented it in specific places, but this is an overall guidance. In many cases the files/components are too big and complicated, they need to be split up.

I have a warning in console. Please, fix it.
Screenshot 2022-08-04 at 16 49 58

The token & amount are of different font/alignment (in figma they are the same)
Screenshot 2022-08-04 at 16 51 57

Please, ping me once you finish with the updates - I'll make sure to review with priority (while it's also more fresh in my mind).

File: src/modules/dashboard/components/Dialogs/EditExpenditureDialog/EditExpenditureDialogForm.tsx [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+import styles from './EditExpenditureDialogForm.css';
+
+export const MSG = defineMessages({
+  header: {
+    id: 'dashboard.EditExpenditureDialog.EditExpenditureDialogForm.header',
+    defaultMessage: 'Create a motion to edit payment',
+  },
+  descriptionText: {
+    id: `dashboard.EditExpenditureDialog.EditExpenditureDialogForm.descriptionText`,
+    defaultMessage: `Payment is currently at the locked stage.
```
Comment (arrenv) [inline src/modules/dashboard/components/Dialogs/EditExpenditureDialog/EditExpenditureDialogForm.tsx:32]: Please change this text to be: "This Payment is currently Locked. Either a Motion, or a member with the Arbitration permission are required to make changes. "

**Comment by arrenv** [comment]
@agnieszkajarosikloj Regarding the Tooltip, which I seemed to have missed the notification for. Please use this copy:

> There is an active Motion to change this payment.

As of the design here- https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=5675%3A322900

**Review by arrenv — CHANGES_REQUESTED** [review]
Thank you @agnieszkajarosikloj, this is certainly a big one. Some feedback:

1. This padding is uneven in the modal

37d84219-a4dc-4837-b445-dc1582c53cdf

2. These are inconsistent, please use this format for both: `1: @username, 100 TKN, 1 hours`

download

download (1)


3. I added a new recipient and then discarded it in the modal and I received the following message, also nothing happened when click the "Discard" button on the 1st recipient.

863219b5-6975-416e-b4df-0a53fe95c744

4. Could we please have consistent fonts for each of these:

download (2)

Please have them all match this font:

```
font-size: var(--size-normal);
font-weight: 700;
 ```

5. Please update this message in yellow to match the design:

8dbda23a-06b2-4d1d-b74b-554dfd67c3bd

Screenshot 2022-08-08 165449

**Review by chinins — CHANGES_REQUESTED** [review]
Much better - easier to read the code and understand all the parts.
On top of my direct comments, I've noticed the following:

1. The css warning is still there
Screenshot 2022-08-08 at 18 43 47



2. There is a `key` warning in the edit dialog - I guess because you use id as `key` and it's the same users that are being changed. Also claim delay for the second user shouldn't be there as it didn't change/wasn't added.

Screenshot 2022-08-08 at 18 49 00

3. I encounter a bug (similar to Arren's) when I click on discard of the second user - just claim delay disappears and all values switched to be the same for both items
Screenshot 2022-08-08 at 18 51 07

This is what happens when I click discard second time 
Screenshot 2022-08-08 at 18 51 23

**Comment by Julianahlk** [comment]
> @Julianahlk thank you for comments.

> 1. I've fixed 'Completed' button color in this PR: https://github.com/JoinColony/colonyDapp/pull/3527/files#diff-692e723973e10e96eea291e621374752d4483680460c120175c41afed277eef3
>    I don't want to copy the same code here.
> 2. It is similar to the first point. Claim delay [PR](https://github.com/JoinColony/colonyDapp/pull/3527) isn't merged yet to the [advanced-expenditures](https://github.com/JoinColony/colonyDapp/tree/feature/advanced-expenditures) branch, so delay doesn't work yet.
> 3. and 4. I've added fixes. Thank you for pointing this out.
> 4. I couldn't find tooltip for this dot in the desing. @arrenv should tooltip be here?  If so, can you please tell me what should be displayed here?

I see the changes you made. Looking good!
Screen Shot 2022-08-11 at 10 01 23 PM
Screen Shot 2022-08-11 at 10 06 40 PM

A few things:
- Can we give the orange tag below a tiny bit more of top and bottom margin space?
Screen Shot 2022-08-11 at 9 52 13 PM

When I pick an username that does not exist I get the following:
Screen Shot 2022-08-11 at 10 21 04 PM

**Comment by Julianahlk** [comment]
Hey @agnieszkajarosikloj I am still seeing the following warnings:
Screen Shot 2022-08-16 at 12 08 40 AM

Screen Shot 2022-08-16 at 12 09 43 AM

But things are looking great! Good work 💯🙌✨
Screen Shot 2022-08-16 at 12 18 17 AM

**Review by arrenv — CHANGES_REQUESTED** [review]
Thank you @agnieszkajarosikloj, I am still seeing these issues.

2. These are inconsistent, please use this format for both: `1: @username, 100 TKN, 1 hours`

download

download (1)

4. Could we please have consistent fonts for each of these:

download (2)

Please have them all match this font:

```
font-size: var(--size-normal);
font-weight: 700;
 ```

I also received the same unique key prop error as @Julianahlk, however, I don't seem to be able to replicate it too easily.

File: src/modules/dashboard/components/ExpenditurePage/Payments/LockedPayments.tsx [hunk 7]
```diff
@@ -69,75 +77,91 @@ const LockedPayments = ({
...
-            recipient.recipient.profile.displayName ||
-            recipient.recipient.profile.username ||
-            '';
+
+          const inPendingState = pendingChanges?.recipients?.find(
+            (item) => item.id === recipient.id,
+          );
 
           return (
             <div className={styles.singleRecipient} key={recipient.id}>
```
Comment (arrenv) [inline src/modules/dashboard/components/ExpenditurePage/Payments/LockedPayments.tsx:109]: This is a key that is causing errors for me when I use the same recipient for multiple payments. I think we could use the count as well as the recipient.id i.e. address as the key. For example `0xaEsd0...wE73_1`.

31620875-2a03860425fd52af46b745d2222db8e3

File: src/modules/dashboard/components/ExpenditurePage/Payments/LockedPayments.tsx [hunk 8]
```diff
@@ -69,75 +77,91 @@ const LockedPayments = ({
...
+                </>
+              )}
+            </div>
+          );
+        })}
+      </div>
+      <div className={styles.recipientContainer}>
+        {pendingCreations?.map((newRecipient, index) => {
+          return (
+            <div className={styles.singleRecipient} key={newRecipient.id}>
```
Comment (arrenv) [inline src/modules/dashboard/components/ExpenditurePage/Payments/LockedPayments.tsx:149]: Same here, please use the count as well as the recipient.id i.e. address as the key. For example `0xaEsd0...wE73_1`

File: src/modules/dashboard/components/Dialogs/EditExpenditureDialog/ChnagedRecipients/ChangedRecipients.tsx [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+                                    oldValue.recipient?.profile.displayName ||
+                                    ''
+                                  }
+                                />
+                              ),
+                              value: recipientValues?.map(
+                                ({ amount, token }) =>
+                                  token &&
+                                  amount && (
+                                    <div key={token.id}>
```
Comment (arrenv) [inline src/modules/dashboard/components/Dialogs/EditExpenditureDialog/ChnagedRecipients/ChangedRecipients.tsx:164]: This seems to be causing an issue when the same token is added twice.

File: src/modules/dashboard/components/ExpenditurePage/Payments/RecipientHeader/RecipientHeader.tsx [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+                        recipient.recipient?.profile.displayName ||
+                        ''
+                      }
+                    />
+                  ),
+                  value: recipientValues?.map(
+                    ({ amount, token }) =>
+                      token &&
+                      amount && (
+                        <div className={styles.tokenWrapper} key={token.id}>
```
Comment (arrenv) [inline src/modules/dashboard/components/ExpenditurePage/Payments/RecipientHeader/RecipientHeader.tsx:94]: Would this cause key issues if you add the same token twice?

**Review by arrenv — CHANGES_REQUESTED** [review]
@agnieszkajarosikloj Thank you, the updates look good. I think I identified the unique "key" issues, please check my comments.

**Review by arrenv — APPROVED** [review]
Lots of work on this, looking good to me, thank you @agnieszkajarosikloj

**Review by chinins — APPROVED** [review]
That is a lot of work - well done!

**Review by Julianahlk — APPROVED** [review]
Great work @agnieszkajarosikloj 🙌

**Pull request merged**